cluster-autoscaler/cloudprovider/gce/templates.go

```diff
@@ -228,18 +231,61 @@ func (t *GceTemplateBuilder) BuildNodeFromTemplate(mig Mig, template *gce.Instan
 	return &node, nil
 }
 
-// isEphemeralStorageWithInstanceTemplateDisabled will allow bypassing Disk Size of Boot Disk from being
+func ephemeralStorageLocalSSDCount(kubeEnvValue string) int64 {
+	v, found, err := extractAutoscalerVarFromKubeEnv(kubeEnvValue, "EPH_STORAGE_LOCAL_SSD_COUNT")
+	if err != nil {
+		klog.Warningf("cannot extract EPH_STORAGE_LOCAL_SSD_COUNT from kube-env, default to 0: %v", err)
+		return 0
+	}
+
+	if !found {
+		return 0
+	}
+
+	n, err := strconv.Atoi(v)
+	if err != nil {
+		klog.Warningf("cannot parse EPH_STORAGE_LOCAL_SSD_COUNT value, default to 0: %v", err)
+		return 0
+	}
+
+	return int64(n)
+}
+
+func getLocalSSDEphemeralStorageFromInstanceTemplateProperties(instanceProperties *gce.InstanceProperties, ssdCount int64) (ephemeralStorage int64, err error) {
+	if instanceProperties.Disks == nil {
+		return 0, fmt.Errorf("instance properties disks is nil")
+	}
+
+	var count int64
+	for _, disk := range instanceProperties.Disks {
+		if disk != nil && disk.InitializeParams != nil {
+			if disk.Type == "SCRATCH" && disk.InitializeParams.DiskType == "local-ssd" {
+				// we could just multiply ssdCount by a constant 375GiB, because all GCE local SSDs are 375GiB
+				// but this loop verifies that the instance template has at least the requested number of local SSDs
+				// and is forward-compatible with a potentially different size
```

Comment:
> This only works for the case when a new disk size is introduced and all the disks for the node are of the same size.

That's what I had in mind, e.g., imagine that a new machine family is introduced with 500GiB local SSDs. It wouldn't make much sense to have local SSDs of different sizes on the same node IMO, because that's not RAID-friendly, but to avoid unsuspecting bugs, I could continue the loop to check that all disks are the same size. Wdyt?

Reply:
We should at least make sure that enough local SSDs are attached. A loop is needed to count them. So we might as well also check that all disks are the same size, and use that size.